node/src/types/block.rs

```diff
@@ -1111,6 +1113,209 @@ impl Item for BlockHeaderWithMetadata {
+        verifiable_chunked_hash_activation: EraId,
+    ) -> Result<BlockHeader, BlockHeadersBatchValidationError> {
+        let highest_header = self
+            .0
+            .first()
+            .ok_or(BlockHeadersBatchValidationError::BatchEmpty)?;
+
+        // Check first header first b/c it's cheaper than verifying continuity.
+        let highest_hash = highest_header.hash(verifiable_chunked_hash_activation);
+        if &highest_hash != latest_known.parent_hash() {
+            return Err(BlockHeadersBatchValidationError::HighestBlockHashMismatch {
+                expected: *latest_known.parent_hash(),
+                got: highest_hash,
+            });
+        }
+
+        if !Self::is_continuous_and_descending(self.inner(), verifiable_chunked_hash_activation) {
+            return Err(BlockHeadersBatchValidationError::BatchNotContinuous);
+        }
+
+        self.0
+            .last()
+            .cloned()
+            .ok_or(BlockHeadersBatchValidationError::BatchEmpty)
+    }
+
+    /// Tries to create an instance of `BlockHeadersBatch` from a `Vec<BlockHeader>`.
+    ///
+    /// Returns `Some(Self)` if data passes validation, otherwise `None`.
+    pub fn from_vec(
+        batch: Vec<BlockHeader>,
+        requested_id: &BlockHeadersBatchId,
+        verifiable_chunked_hash_activation: EraId,
+    ) -> Option<Self> {
+        match batch.first() {
+            Some(highest) => {
+                if highest.height() != requested_id.highest {
+                    error!(expected_highest=?requested_id.highest, got_highest=?highest, "unexpected highest block header");
+                    return None;
+                }
+            }
+            None => {
+                error!("response cannot be an empty batch");
+                return None;
+            }
+        }
+
+        match batch.last() {
+            Some(lowest) => {
+                if lowest.height() != requested_id.lowest {
+                    error!(expected_lowest=?requested_id.lowest, got_lowest=?lowest, "unexpected lowest block header");
+                    return None;
+                }
+            }
+            None => {
+                error!("input cannot be empty");
+                return None;
+            }
+        }
+
+        if !Self::is_continuous_and_descending(&batch, verifiable_chunked_hash_activation) {
+            error!("batch is not continuous");
+            return None;
+        }
+
+        Some(Self(batch))
+    }
+
+    /// Returns inner value.
+    pub fn into_inner(self) -> Vec<BlockHeader> {
+        self.0
+    }
+
+    /// Returns a reference to an inner vector of block headers.
+    pub fn inner(&self) -> &Vec<BlockHeader> {
+        &self.0
+    }
+
+    /// Tests whether the block header batch is continuous and in descending order.
+    pub fn is_continuous_and_descending(
+        batch: &[BlockHeader],
+        verifiable_chunked_hash_activation: EraId,
+    ) -> bool {
+        batch
```

Comment:
While I do appreciate functional iterator magic, I think 

```rust
for i in 1..batch.len() {
    let current = batch[i];
    let pred = batch[i-1];

     if current.parent_hash() != pred.hash(...) { ... }
     if curent.height() != pred.height() +1 { ... }
}
```

is maybe a little easier to read and review? :)

Reply:
I'm going to resolve the comment for now :)